compiler/rmi/kotlinr/src/org/jetbrains/kotlin/rmi/kotlinr/KotlinCompilerClient.kt

```diff
@@ -18,246 +18,376 @@ package org.jetbrains.kotlin.rmi.kotlinr
+        }
+        finally {
+            // assuming that all important output is already done, the rest should be routed to the log by the daemon itself
+            if (stdoutThread.isAlive) {
+                // TODO: find better method to stop the thread, but seems it will require asynchronous consuming of the stream
+                stdoutThread.stop()
             }
         }
+    }
 
-        data class ClientOptions(
-                public var stop: Boolean = false
-        ) : OptionsGroup {
-            override val mappers: List<PropMapper<*, *, *>>
-                get() = listOf( BoolPropMapper(this, ::stop))
-        }
 
-        jvmStatic public fun main(vararg args: String) {
-            val compilerId = CompilerId()
-            val daemonOptions = DaemonOptions()
-            val daemonLaunchingOptions = DaemonJVMOptions()
-            val clientOptions = ClientOptions()
-            val filteredArgs = args.asIterable().filterExtractProps(compilerId, daemonOptions, daemonLaunchingOptions, clientOptions, prefix = COMPILE_DAEMON_CMDLINE_OPTIONS_PREFIX)
-
-            if (!clientOptions.stop) {
-                if (compilerId.compilerClasspath.none()) {
-                    // attempt to find compiler to use
-                    println("compiler wasn't explicitly specified, attempt to find appropriate jar")
-                    System.getProperty("java.class.path")
-                            ?.split(File.pathSeparator)
-                            ?.map { File(it).parent }
-                            ?.distinct()
-                            ?.map {
-                                it?.walk()
-                                        ?.firstOrNull { it.getName().equals(COMPILER_JAR_NAME, ignoreCase = true) }
-                            }
-                            ?.filterNotNull()
-                            ?.firstOrNull()
-                            ?.let { compilerId.compilerClasspath = listOf(it.absolutePath) }
-                }
-                if (compilerId.compilerClasspath.none())
-                    throw IllegalArgumentException("Cannot find compiler jar")
-                else
-                    println("desired compiler classpath: " + compilerId.compilerClasspath.joinToString(File.pathSeparator))
+    private fun checkCompilerId(compiler: CompileService, localId: CompilerId): Boolean {
+        val remoteId = compiler.getCompilerId()
+        return (localId.compilerVersion.isEmpty() || localId.compilerVersion == remoteId.compilerVersion) &&
+               (localId.compilerClasspath.all { remoteId.compilerClasspath.contains(it) }) &&
+               (localId.compilerDigest.isEmpty() || remoteId.compilerDigest.isEmpty() || localId.compilerDigest == remoteId.compilerDigest)
+    }
 
-                compilerId.updateDigest()
+
+    class FileBasedLock(compilerId: CompilerId, daemonOptions: DaemonOptions) {
+        private val lockFile: File =
+                File(daemonOptions.runFilesPath,
```

Comment:
Taking into account that I need recognisable file names and use parts of the names for discovery, it seems that straightforward approach with temdir has several potential security flaws, to fix which I'd need a platform-specific code with native function calls (at least until an access to a robust way to create a secure temp subdirs, implemented in java 7).
Another problem is that e.g. Idea substitutes "java.io.tmpdir", so it is no longer points to the system one, which with current mechanism makes sharing daemons between various clients impossible.
Homedir solution probably is not ideal, but seems safe enough, and there is a fallback possibility to specify run directory for daemons.